// opencode-auto-resume/dist/index.js
// @bun
// src/index.ts
var DEFAULT_CHUNK_TIMEOUT_MS = 45000;
var DEFAULT_CHECK_INTERVAL_MS = 5000;
var DEFAULT_GRACE_PERIOD_MS = 3000;
var DEFAULT_MAX_RETRIES = 3;
var DEFAULT_MAX_BACKOFF_MS = 8000;
var DEFAULT_BASE_BACKOFF_MS = 1000;
var DEFAULT_SUBAGENT_WAIT_MS = 15000;
var ABORT_CONTINUE_DELAY_MS = 2000;
var DEFAULT_LOOP_MAX_CONTINUES = 3;
var DEFAULT_LOOP_WINDOW_MS = 10 * 60000;
var TOOL_TEXT_CHECK_DELAY_MS = 1500;
var MAX_IDLE_SESSIONS = 50;
var IDLE_CLEANUP_MS = 10 * 60000;
var SESSION_DISCOVERY_INTERVAL_MS = 60000;
var TOOL_TEXT_RECOVERY_PROMPT = "Your last message contained a raw tool call printed as text instead of being executed. " + "Please use the proper tool calling mechanism to execute it.";
var THINKING_TOOL_RECOVERY_PROMPT = "I noticed you have a tool call generated in your thinking/reasoning. " + "Please execute it using the proper tool calling mechanism instead of keeping it in reasoning.";
var TOOL_TEXT_PATTERNS = [
  /<function\s*=/i,
  /<function>/i,
  /<\/function>/i,
  /<parameter\s*=/i,
  /<parameter>/i,
  /<\/parameter>/i,
  /<tool_call[\s>]/i,
  /<\/tool_call>/i,
  /<tool[\s_]name\s*=/i,
  /<invoke\s+/i,
  /<func(?:t|ti|tio|tion)?$/im,
  /<par(?:a|am|ame|amet|amete|ameter)?$/im,
  /<(?:edit|write|read|bash|grep|glob|search|replace|execute|run)\s*(?:\s[^>]*)?\s*(?:\/>|>)/i
];
var TRUNCATED_XML_PATTERNS = [
  { open: /<function[^>]*>/i, close: /<\/function>/i },
  { open: /<parameter[^>]*>/i, close: /<\/parameter>/i },
  { open: /<tool_call[^>]*>/i, close: /<\/tool_call>/i }
];
var READY_TO_CONTINUE_PATTERNS = [
  /ready to continue with task/i,
  /continuing with task/i,
  /continue with task/i,
  /proceeding with task/i,
  /ready to proceed with task/i,
  /will continue with task/i,
  /moving on to task/i
];
function containsToolCallAsText(text) {
  if (text.length <= 10)
    return false;
  if (TOOL_TEXT_PATTERNS.some((pat) => pat.test(text)))
    return true;
  for (const { open, close } of TRUNCATED_XML_PATTERNS) {
    if (open.test(text) && !close.test(text))
      return true;
  }
  return false;
}
function containsReadyToContinuePattern(text) {
  const lines = text.split(`
`);
  const lastLine = lines[lines.length - 1]?.trim();
  if (!lastLine)
    return false;
  const lastLines = lines.slice(-3).join(`
`);
  return READY_TO_CONTINUE_PATTERNS.some((pat) => pat.test(lastLines));
}
var AutoResumePlugin = async (ctx, options) => {
  const chunkTimeoutMs = options?.chunkTimeoutMs ?? DEFAULT_CHUNK_TIMEOUT_MS;
  const checkIntervalMs = options?.checkIntervalMs ?? DEFAULT_CHECK_INTERVAL_MS;
  const gracePeriodMs = options?.gracePeriodMs ?? DEFAULT_GRACE_PERIOD_MS;
  const maxRetries = options?.maxRetries ?? DEFAULT_MAX_RETRIES;
  const maxBackoffMs = options?.maxBackoffMs ?? DEFAULT_MAX_BACKOFF_MS;
  const baseBackoffMs = options?.baseBackoffMs ?? DEFAULT_BASE_BACKOFF_MS;
  const subagentWaitMs = options?.subagentWaitMs ?? DEFAULT_SUBAGENT_WAIT_MS;
  const loopMaxContinues = options?.loopMaxContinues ?? DEFAULT_LOOP_MAX_CONTINUES;
  const loopWindowMs = options?.loopWindowMs ?? DEFAULT_LOOP_WINDOW_MS;
  const sessions = new Map;
  let timer = null;
  let discoveryTimer = null;
  let initialised = false;
  let prevBusyCount = 0;
  function recordContinue(sid) {
    const w = sessions.get(sid);
    if (!w)
      return;
    w.continueTimestamps.push(Date.now());
    const cutoff = Date.now() - loopWindowMs;
    while (w.continueTimestamps.length > 0 && w.continueTimestamps[0] < cutoff) {
      w.continueTimestamps.shift();
    }
  }
  function isHallucinationLoop(sid) {
    const w = sessions.get(sid);
    if (!w)
      return false;
    recordContinue(sid);
    return w.continueTimestamps.length >= loopMaxContinues;
  }
  async function log(level, msg) {
    try {
      await ctx.client.app.log({
        body: { service: "auto-resume", level, message: msg }
      });
    } catch {}
  }
  function ensureWatch(sid) {
    let w = sessions.get(sid);
    if (!w) {
      w = {
        lastActivityAt: Date.now(),
        status: "unknown",
        userCancelled: false,
        resumeAttempts: 0,
        lastRetryAt: 0,
        gaveUp: false,
        orphanWatchStartAt: null,
        aborting: false,
        toolTextRecovered: false,
        toolTextAttempts: 0,
        continueTimestamps: [],
        idleSince: null,
        continuing: false
      };
      sessions.set(sid, w);
    }
    return w;
  }
  function touchSession(sid) {
    const w = sessions.get(sid);
    if (w && w.status === "busy" && !w.userCancelled) {
      w.lastActivityAt = Date.now();
    }
  }
  function busyCount() {
    let count = 0;
    for (const [, w] of sessions) {
      if (w.status === "busy" && !w.userCancelled)
        count++;
    }
    return count;
  }
  function getLoneBusySession() {
    let found = null;
    let count = 0;
    for (const [sid, w] of sessions) {
      if (w.status === "busy" && !w.userCancelled) {
        count++;
        found = { sid, w };
      }
    }
    return count === 1 ? found : null;
  }
  function getSid(ev) {
    const props = ev.properties;
    return ev.sessionID ?? props?.sessionID ?? props?.part?.sessionID ?? props?.info?.sessionID;
  }
  function getError(ev) {
    const props = ev.properties;
    return ev.error ?? props?.error;
  }
  function getStatus(ev) {
    const props = ev.properties;
    return ev.status ?? props?.status;
  }
  function short(sid) {
    return sid.length > 12 ? `...${sid.slice(-8)}` : sid;
  }
  function backoffMs(attempt) {
    return Math.min(baseBackoffMs * Math.pow(2, attempt - 1), maxBackoffMs);
  }
  function cleanupIdleSessions() {
    const now = Date.now();
    const toDelete = [];
    let idleCount = 0;
    for (const [sid, w] of sessions) {
      if (w.status !== "busy") {
        idleCount++;
        if (w.idleSince && now - w.idleSince > IDLE_CLEANUP_MS) {
          toDelete.push(sid);
        }
      }
    }
    if (idleCount > MAX_IDLE_SESSIONS) {
      const idleEntries = [];
      for (const [sid, w] of sessions) {
        if (w.status !== "busy" && w.idleSince) {
          idleEntries.push({ sid, idleSince: w.idleSince });
        }
      }
      idleEntries.sort((a, b) => a.idleSince - b.idleSince);
      const excess = idleCount - MAX_IDLE_SESSIONS;
      for (let i = 0;i < excess && i < idleEntries.length; i++) {
        if (!toDelete.includes(idleEntries[i].sid)) {
          toDelete.push(idleEntries[i].sid);
        }
      }
    }
    for (const sid of toDelete) {
      sessions.delete(sid);
    }
    if (toDelete.length > 0) {
      log("debug", `Cleaned up ${toDelete.length} idle session(s). Map size: ${sessions.size}`);
    }
  }
  async function sendContinuePrompt(sid, text, w) {
    if (w.continuing) {
      await log("debug", `${short(sid)} - continue already in progress, skipping`);
      return;
    }
    w.continuing = true;
    try {
      let agent2;
      let model2;
      const msgResp = await ctx.client.session.messages({ path: { id: sid } });
      const msgs = extractMessages(msgResp);
      for (let i = msgs.length - 1;i >= 0; i--) {
        const msg = msgs[i];
        const role = msg.role ?? msg.info?.role;
        if (role === "user") {
          const rawAgent = msg.agent;
          if (typeof rawAgent === "string") {
            agent2 = rawAgent;
          } else {
            const fallbackAgent = msg.info?.agent;
            if (typeof fallbackAgent === "string") {
              agent2 = fallbackAgent;
            }
          }
          let rawModel = msg.model;
          if (!rawModel) {
            rawModel = msg.info?.model;
          }
          if (rawModel && typeof rawModel.providerID === "string" && typeof rawModel.modelID === "string") {
            model2 = {
              providerID: rawModel.providerID,
              modelID: rawModel.modelID
            };
          }
          break;
        }
      }
      await ctx.client.session.prompt({
        path: { id: sid },
        body: {
          parts: [{ type: "text", text }],
          agent: agent2,
          model: model2
        }
      });
      await log("debug", `${short(sid)} - prompt sent with agent: ${agent2 ?? "(default)"}, model: ${model2 ? `${model2.providerID}/${model2.modelID}` : "(default)"}`);
      recordContinue(sid);
      w.lastRetryAt = Date.now();
    } catch (err) {
      const errMsg = err instanceof Error ? err.message : String(err);
      await log("warn", `${short(sid)} - prompt failed: ${errMsg}`);
      try {
        await ctx.client.session.prompt({
          path: { id: sid },
          body: { parts: [{ type: "text", text }], agent, model }
        });
        recordContinue(sid);
        w.lastRetryAt = Date.now();
      } catch (retryErr) {
        const retryMsg = retryErr instanceof Error ? retryErr.message : String(retryErr);
        await log("error", `${short(sid)} - prompt retry also failed: ${retryMsg}`);
        throw retryErr;
      }
    } finally {
      w.continuing = false;
    }
  }
  function extractMessages(response) {
    if (Array.isArray(response))
      return response;
    if (Array.isArray(response.data))
      return response.data;
    if (Array.isArray(response.messages))
      return response.messages;
    return [];
  }
  const SUBAGENT_STUCK_MS = 60000;
  const SUBAGENT_RECOVERY_PROMPT = "It looks like you may have stalled or timed out. Please retry the last operation or continue with the task.";
  async function recoverSubagent(subagentSid) {
    try {
      await ctx.client.session.prompt({
        path: { id: subagentSid },
        body: { parts: [{ type: "text", text: SUBAGENT_RECOVERY_PROMPT }] }
      });
      await log("info", `Sent recovery prompt to subagent ${short(subagentSid)}`);
      return true;
    } catch (err) {
      const errMsg = err instanceof Error ? err.message : String(err);
      await log("warn", `Failed to recover subagent ${short(subagentSid)}: ${errMsg}`);
      return false;
    }
  }
  async function checkSubagentStatus(parentSid) {
    try {
      const response = await ctx.client.session.list();
      const allSessions = extractMessages(response);
      const now = Date.now();
      let hasBusySubagent = false;
      for (const s of allSessions) {
        const sId = s.id;
        if (!sId || sId === parentSid)
          continue;
        const status = s.status;
        if (status === "busy") {
          hasBusySubagent = true;
          const msgResponse = await ctx.client.session.messages({ path: { id: sId } });
          const messages = extractMessages(msgResponse);
          const lastMsg = messages[messages.length - 1];
          const rawRole = lastMsg?.role ?? lastMsg?.info?.role;
          if (lastMsg && rawRole === "assistant" && (("error" in lastMsg) || lastMsg.info && ("error" in lastMsg.info))) {
            await log("debug", `Subagent ${short(sId)} appears crashed`);
            return "crashed";
          }
          const msgTime = lastMsg?.time?.created ?? lastMsg?.time;
          const hasToolCall = lastMsg?.toolCall !== undefined || lastMsg?.tool_calls !== undefined || lastMsg?.parts?.some((p) => p.type === "tool-call") !== undefined;
          const isStuck = hasToolCall ? now - msgTime > SUBAGENT_STUCK_MS * 3 : now - msgTime > SUBAGENT_STUCK_MS;
          if (isStuck) {
            await log("debug", `Subagent ${short(sId)} stuck - no new text in >${hasToolCall ? 3 : 1}min`);
            return { status: "crashed", stuckSid: sId };
          }
        }
      }
      if (!hasBusySubagent) {
        return { status: "idle" };
      }
      return { status: "busy" };
    } catch (err) {
      const errMsg = err instanceof Error ? err.message : String(err);
      await log("debug", `checkSubagentStatus failed: ${errMsg}`);
      return { status: "unknown" };
    }
  }
  function resetSessionFlags(w) {
    w.userCancelled = false;
    w.resumeAttempts = 0;
    w.gaveUp = false;
    w.orphanWatchStartAt = null;
    w.aborting = false;
    w.toolTextRecovered = false;
    w.toolTextAttempts = 0;
    w.continueTimestamps = [];
    w.idleSince = null;
    w.continuing = false;
  }
  function resetIdleFlags(w) {
    w.userCancelled = false;
    w.aborting = false;
    w.orphanWatchStartAt = null;
    w.idleSince = Date.now();
  }
  async function checkForToolCallAsText(sid, w) {
    if (typeof sid !== "string" || !sid)
      return;
    if (w.userCancelled || w.toolTextRecovered)
      return;
    if (w.toolTextAttempts > 0) {
      const elapsed = Date.now() - w.lastRetryAt;
      const requiredBackoff = backoffMs(w.toolTextAttempts);
      if (elapsed < requiredBackoff)
        return;
    }
    if (w.toolTextAttempts >= maxRetries)
      return;
    await log("debug", `${short(sid)} - checking for tool-call-as-text (attempt ${w.toolTextAttempts + 1})`);
    try {
      const response = await ctx.client.session.messages({
        path: { id: sid }
      });
      const messages = extractMessages(response);
      const recent = messages.slice(-3);
      let bestCandidate = null;
      for (const msg of recent) {
        const rawRole = msg.role ?? msg.info?.role;
        if (rawRole !== "assistant")
          continue;
        const parts = msg.parts;
        if (!parts)
          continue;
        for (const part of parts) {
          const partType = part.type;
          let text = "";
          let isReasoning = false;
          if (partType === "text") {
            text = part.text ?? "";
          } else if (partType === "reasoning") {
            text = part.text ?? "";
            isReasoning = true;
          } else {
            continue;
          }
          if (containsToolCallAsText(text)) {
            const candidate = {
              prompt: isReasoning ? THINKING_TOOL_RECOVERY_PROMPT : TOOL_TEXT_RECOVERY_PROMPT,
              source: isReasoning ? "reasoning" : "text",
              priority: 0
            };
            if (!bestCandidate || candidate.priority < bestCandidate.priority) {
              bestCandidate = candidate;
            }
          }
          if (containsReadyToContinuePattern(text)) {
            const candidate = {
              prompt: "continue",
              source: "ready-to-continue",
              priority: 1
            };
            if (!bestCandidate || candidate.priority < bestCandidate.priority) {
              bestCandidate = candidate;
            }
          }
        }
      }
      if (!bestCandidate)
        return;
      w.toolTextRecovered = true;
      w.toolTextAttempts++;
      await log("info", `${bestCandidate.source} detected on ${short(sid)}! ` + `Attempt ${w.toolTextAttempts}/${maxRetries}. Sending recovery prompt...`);
      if (isHallucinationLoop(sid)) {
        await log("warn", `Hallucination loop detected on ${short(sid)} \u2014 aborting instead`);
        await tryAbortAndResume(sid, w);
      } else {
        try {
          await sendContinuePrompt(sid, bestCandidate.prompt, w);
          await log("info", `${short(sid)} - ${bestCandidate.source} recovery sent (attempt ${w.toolTextAttempts})`);
        } catch (err) {
          const errMsg = err instanceof Error ? err.message : String(err);
          await log("warn", `${short(sid)} - ${bestCandidate.source} recovery failed: ${errMsg}`);
        }
      }
    } catch (err) {
      const errMsg = err instanceof Error ? err.message : String(err);
      log("debug", `${short(sid)} - could not fetch messages: ${errMsg}`);
    }
  }
  async function tryAbortAndResume(sid, w) {
    if (typeof sid !== "string" || !sid)
      return false;
    if (w.aborting)
      return false;
    w.aborting = true;
    const idleSec = Math.round((Date.now() - (w.orphanWatchStartAt ?? w.lastActivityAt)) / 1000);
    await log("info", `Abort+Resume on ${short(sid)} (${idleSec}s idle). Aborting...`);
    try {
      await ctx.client.session.abort({ sessionID: sid });
      await log("info", `${short(sid)} - abort OK`);
    } catch (err) {
      const errMsg = err instanceof Error ? err.message : String(err);
      await log("warn", `${short(sid)} - abort failed: ${errMsg}`);
      w.aborting = false;
      return false;
    }
    await new Promise((resolve) => setTimeout(resolve, ABORT_CONTINUE_DELAY_MS));
    if (w.status === "busy")
      w.status = "idle";
    try {
      await sendContinuePrompt(sid, "continue", w);
      await log("info", `${short(sid)} - abort+continue done`);
      w.orphanWatchStartAt = null;
      w.resumeAttempts++;
      w.aborting = false;
      return true;
    } catch (err) {
      const errMsg = err instanceof Error ? err.message : String(err);
      await log("warn", `${short(sid)} - continue after abort failed: ${errMsg}`);
      w.aborting = false;
      return false;
    }
  }
  async function tryResume(sid, w, reason) {
    if (typeof sid !== "string" || !sid) {
      await log("warn", `tryResume called with invalid sid: ${sid}`);
      return false;
    }
    const now = Date.now();
    const elapsedSinceRetry = now - w.lastRetryAt;
    const requiredBackoff = backoffMs(w.resumeAttempts);
    if (w.lastRetryAt > 0 && elapsedSinceRetry < requiredBackoff)
      return false;
    if (isHallucinationLoop(sid)) {
      await log("warn", `Hallucination loop on ${short(sid)}! Aborting...`);
      return await tryAbortAndResume(sid, w);
    }
    w.resumeAttempts++;
    const idleSec = Math.round((now - w.lastActivityAt) / 1000);
    await log("info", `${reason} on ${short(sid)} (${idleSec}s, retry ${w.resumeAttempts}/${maxRetries})`);
    try {
      await sendContinuePrompt(sid, "continue", w);
      await log("info", `${short(sid)} - retry sent`);
      return true;
    } catch (err) {
      const errMsg = err instanceof Error ? err.message : String(err);
      await log("warn", `${short(sid)} - retry failed: ${errMsg}`);
      w.lastRetryAt = now;
      return false;
    }
  }
  async function discoverSessions() {
    try {
      const response = await ctx.client.session.list();
      const list = extractMessages(response);
      for (const s of list) {
        const sid = s.id;
        if (sid) {
          const isNew = !sessions.has(sid);
          ensureWatch(sid);
          const status = s.status;
          if (status) {
            const w = sessions.get(sid);
            w.status = status;
            if (status === "idle")
              w.idleSince = Date.now();
          }
          if (isNew) {
            log("debug", `Discovered session ${short(sid)} via list()`);
          }
        }
      }
    } catch (err) {
      const errMsg = err instanceof Error ? err.message : String(err);
      log("debug", `Session discovery failed: ${errMsg}`);
    }
  }
  function startTimer() {
    if (timer)
      return;
    timer = setInterval(async () => {
      const now = Date.now();
      const numBusy = busyCount();
      for (const [sid, w] of sessions) {
        if (w.status !== "busy")
          continue;
        if (w.userCancelled)
          continue;
        if (w.aborting)
          continue;
        if (w.orphanWatchStartAt !== null) {
          const orphanIdle = now - w.orphanWatchStartAt;
          if (orphanIdle >= subagentWaitMs + gracePeriodMs) {
            if (w.resumeAttempts < maxRetries) {
              const subStatus = await checkSubagentStatus(sid);
              if (subStatus.status === "crashed" && subStatus.stuckSid) {
                const recovered = await recoverSubagent(subStatus.stuckSid);
                if (recovered) {
                  await log("info", `Sent recovery prompt to stuck subagent ${short(subStatus.stuckSid)}, waiting...`);
                } else {
                  await log("info", `Subagent crashed, triggering abort+resume on ${short(sid)}`);
                  tryAbortAndResume(sid, w);
                }
              } else if (subStatus.status === "idle") {
                await log("info", `All subagents idle, parent ${short(sid)} stuck. Triggering abort+resume.`);
                tryAbortAndResume(sid, w);
              } else {
                await log("debug", `Subagent still running, waiting...`);
              }
            } else if (!w.gaveUp) {
              w.gaveUp = true;
              w.orphanWatchStartAt = null;
              w.aborting = false;
              log("warn", `${short(sid)} - orphan retries exhausted.`);
            }
          }
          continue;
        }
        if (numBusy > 1)
          continue;
        if (w.lastActivityAt > 0 && now - w.lastActivityAt > subagentWaitMs) {
          const subStatus = await checkSubagentStatus(sid);
          if (subStatus.status === "idle" || subStatus.status === "unknown") {
            await log("info", `Parent ${short(sid)} stuck with no active subagents. Triggering abort+resume.`);
            tryAbortAndResume(sid, w);
            continue;
          } else if (subStatus.status === "crashed" && subStatus.stuckSid) {
            const recovered = await recoverSubagent(subStatus.stuckSid);
            if (recovered) {
              await log("info", `Sent recovery prompt to stuck subagent ${short(subStatus.stuckSid)}, waiting...`);
            } else {
              await log("info", `Parent ${short(sid)} subagent recovery failed. Triggering abort+resume.`);
              tryAbortAndResume(sid, w);
            }
            continue;
          }
        }
        const idle = now - w.lastActivityAt;
        if (idle >= chunkTimeoutMs + gracePeriodMs) {
          if (w.resumeAttempts < maxRetries) {
            tryResume(sid, w, "Stream stall");
          } else if (!w.gaveUp) {
            w.gaveUp = true;
            log("warn", `${short(sid)} - all ${maxRetries} retries exhausted.`);
          }
        }
      }
      cleanupIdleSessions();
    }, checkIntervalMs);
    if (timer.unref)
      timer.unref();
    discoveryTimer = setInterval(() => {
      discoverSessions();
    }, SESSION_DISCOVERY_INTERVAL_MS);
    if (discoveryTimer.unref)
      discoveryTimer.unref();
    setTimeout(discoverSessions, 5000);
  }
  startTimer();
  function handleEvent(ev) {
    const type = ev.type;
    const sid = getSid(ev);
    if (sid) {
      touchSession(sid);
    }
    switch (type) {
      case "session.status": {
        if (!sid)
          break;
        const status = getStatus(ev);
        const statusType = status?.type ?? "unknown";
        const w = ensureWatch(sid);
        w.status = statusType;
        if (statusType === "busy") {
          w.lastActivityAt = Date.now();
          resetSessionFlags(w);
          log("debug", `${short(sid)} -> busy (${busyCount()})`);
        } else if (statusType === "idle") {
          w.status = "idle";
          resetIdleFlags(w);
          const currentBusy = busyCount();
          if (prevBusyCount > 1 && currentBusy === 1) {
            const lone = getLoneBusySession();
            if (lone && lone.w.orphanWatchStartAt === null) {
              lone.w.orphanWatchStartAt = Date.now();
              log("info", `Subagent finished, parent ${short(lone.sid)} stuck. Orphan watch (${subagentWaitMs / 1000}s).`);
            }
          }
          prevBusyCount = currentBusy;
          log("debug", `${short(sid)} -> idle (${currentBusy})`);
          if (!w.toolTextRecovered && w.toolTextAttempts < maxRetries) {
            setTimeout(() => {
              checkForToolCallAsText(sid, w);
            }, TOOL_TEXT_CHECK_DELAY_MS);
          }
        } else if (statusType === "retry") {
          touchSession(sid);
          log("debug", `${short(sid)} -> provider retry`);
        }
        break;
      }
      case "session.created": {
        if (!sid)
          break;
        ensureWatch(sid);
        log("debug", `New session: ${short(sid)} (${sessions.size})`);
        break;
      }
      case "session.updated": {
        if (sid)
          ensureWatch(sid);
        break;
      }
      case "session.idle": {
        if (!sid)
          break;
        const w = sessions.get(sid);
        if (w) {
          w.status = "idle";
          resetIdleFlags(w);
          if (!w.toolTextRecovered && w.toolTextAttempts < maxRetries) {
            setTimeout(() => {
              checkForToolCallAsText(sid, w);
            }, TOOL_TEXT_CHECK_DELAY_MS);
          }
        }
        break;
      }
      case "session.error": {
        const errorObj = getError(ev);
        const errorName = errorObj?.name ?? "";
        const isMessageAborted = errorName === "MessageAbortedError";
        if (isMessageAborted) {
          for (const [wSid, w] of sessions) {
            if (w.status === "busy") {
              w.userCancelled = true;
              w.status = "idle";
              resetIdleFlags(w);
            }
          }
          log("info", "User abort (ESC)");
          break;
        }
        if (busyCount() === 0)
          break;
        const errorMessage = errorObj?.data?.message ?? String(errorObj?.data ?? "");
        log("debug", `Session error: ${errorName} - ${errorMessage}`);
        break;
      }
      case "command.executed": {
        for (const [, w] of sessions) {
          resetSessionFlags(w);
        }
        break;
      }
    }
  }
  return {
    event: async ({ event }) => {
      if (!initialised) {
        initialised = true;
        log("info", `opencode-auto-resume ready. timeout=${chunkTimeoutMs}ms, orphan=${subagentWaitMs}ms, loop=${loopMaxContinues}x/${loopWindowMs / 1000}s`);
      }
      handleEvent(event);
    },
    config: async () => {
      log("info", `opencode-auto-resume config OK`);
      if (ctx.ui && typeof ctx.ui.toast === "function") {
        ctx.ui.toast({
          title: "Auto-Resume Plugin",
          message: `Loaded with ${chunkTimeoutMs}ms timeout, ${loopMaxContinues} loop attempts`,
          variant: "success",
          duration: 5000
        });
      }
    }
  };
};
var src_default = AutoResumePlugin;
export {
  src_default as default,
  AutoResumePlugin
};
